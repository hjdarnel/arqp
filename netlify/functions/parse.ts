import { cabrilloToObject } from 'cabrillo';
import { writeToBuffer } from '@fast-csv/format';
import { Handler } from '@netlify/functions';
import Busboy from 'busboy';
import { PrismaClient } from '@prisma/client';
import { PrismaClientKnownRequestError } from '@prisma/client/runtime';
const prisma = new PrismaClient();

interface Submission {
  file?: {
    content: string;
    filename: {
      filename: string;
    };
  };
  callsign?: string;
  email?: string;
  power?: string;
  assistance?: string;
  multipleOperators?: string;
  contestId?: string;
}

const parseMultipartForm = (event): Promise<Submission> => {
  return new Promise((resolve) => {
    const fields = {};

    const busboy = Busboy({
      headers: event.headers
    });

    busboy.on('file', (fieldname, filestream, filename, _, mimeType) => {
      filestream.on('data', (data) => {
        fields[fieldname] = {
          content: data,
          filename,
          type: mimeType
        };
      });
    });

    busboy.on('field', (fieldName, value) => {
      fields[fieldName] = value;
    });

    busboy.on('finish', () => {
      resolve(fields);
    });

    busboy.write(Buffer.from(event.body, 'base64').toString('utf8'));
  });
};

const handler: Handler = async (event, context) => {
  try {
    if (event.httpMethod !== 'POST') return { statusCode: 404 };
    if (!event.body) throw new Error('Missing request body');

    const body = await parseMultipartForm(event);
    if (!body?.file) throw new Error('Missing required file');
    console.log('Received request', { ...body, file: 'omitted' });

    const cabrillo = Buffer.from(body.file.content, 'base64').toString();
    let parsed = cabrilloToObject(cabrillo, { contest: 'NAQP' });

    Object.entries(parsed).map(([key, value]) => {
      if (key === 'QSO') return (parsed[key] = value);
      if (typeof value === 'string')
        return (parsed[key] = value.replace('\r', ''));
      return (parsed[key] = value);
    });

    if (!parsed?.QSO || parsed?.QSO?.length === 0) {
      throw new Error('Unable to parse QSO records from body');
    }
    console.log(`Parsed ${parsed.QSO.length} QSO records`);
    console.info({ ...parsed, QSO: 'omitted' });

    const created = await prisma.submission.create({
      data: {
        contestId: body.contestId,
        email: body.email,
        callsign: body.callsign,
        power: body.power,
        multipleOperators: body.multipleOperators === 'true',
        assistance: body.assistance === 'true',
        name: parsed.NAME,
        club: parsed.CLUB,
        claimedScore: Number.parseInt(parsed['CLAIMED-SCORE']),
        location: parsed.LOCATION,
        logEmail: parsed.EMAIL,
        logOperator: parsed['CATEGORY-OPERATOR'].split('-')[0],
        logStation: parsed['CATEGORY-STATION'],
        logTransmitter: parsed['CATEGORY-TRANSMITTER'],
        logPower: parsed['CATEGORY-POWER'],
        logAssisted: parsed['CATEGORY-ASSISTED'] === 'ASSISTED',
        logBand: parsed['CATEGORY-BAND'],
        logMode: parsed['CATEGORY-MODE']
      }
    });

    return {
      statusCode: 200,
      body: JSON.stringify(created)
    };
  } catch (err: any) {
    console.error(err.message);
    let message = err.message;
    let statusCode = 500;

    if (process.env.NODE_ENV !== 'development') {
      if (err.message.includes('Unique constraint failed on the constraint')) {
        message =
          'Invalid submission, only one submission allowed per callsign per contest.';
        statusCode = 400;
      } else {
        message = 'Error submitting results.';
      }
    }

    return {
      statusCode,
      body: JSON.stringify({ error: { message } })
    };
  }
};

export { handler };
