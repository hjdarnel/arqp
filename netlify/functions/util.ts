import { cabrilloToObject } from 'cabrillo';
import { AuditLog, Contest } from '@prisma/client';

export const parse = async (
  body,
  currentContest: Contest
): Promise<Omit<AuditLog, 'id' | 'created_at'>> => {
  let parsed;
  try {
    const cabrillo = Buffer.from(body.file.content, 'base64').toString();
    parsed = cabrilloToObject(cabrillo, { contest: 'NAQP' });
  } catch (err) {
    throw new Error('Cannot parse file as Cabrillo log. Is it a valid log?');
  }

  Object.entries(parsed).map(([key, value]) => {
    if (key === 'QSO') return (parsed[key] = value);
    if (typeof value === 'string')
      return (parsed[key] = value.replace('\r', ''));
    return (parsed[key] = value);
  });

  if (
    (!parsed?.QSO || parsed?.QSO?.length === 0) &&
    (!parsed?.Qso || parsed?.Qso?.length === 0)
  ) {
    throw new Error('Unable to parse QSO records from body');
  }
  console.log(`Parsed ${parsed.QSO?.length ?? parsed.Qso.length} QSO records`);
  console.info({ ...parsed, QSO: 'omitted' });

  return {
    contestId: currentContest.id,
    email: body.email,
    callsign: body.callsign,
    category: body.category,
    contestLocation: body.location,
    multipleOperators: body.multipleOperators === 'true',
    assistance: body.assistance === 'true',
    name: parsed.NAME,
    club: parsed.CLUB,
    claimedScore: Number.parseInt(body.claimedScore),
    logScore:
      Number.parseInt(parsed['CLAIMED-SCORE']) ??
      Number.parseInt(parsed['claimed-score']),
    logLocation: parsed.LOCATION,
    logEmail: parsed.EMAIL,
    logOperator:
      parsed['CATEGORY-OPERATOR']?.split('-')[0].toUpperCase() ??
      parsed['Category-Operator'].split('-')[0].toUpperCase(),
    logStation:
      parsed['CATEGORY-STATION']?.toUpperCase() ??
      parsed['Category-Station']?.toUpperCase(),
    logTransmitter:
      parsed['CATEGORY-TRANSMITTER']?.toUpperCase() ??
      parsed['Category-Transmitter']?.toUpperCase(),
    logPower:
      parsed['CATEGORY-POWER']?.toUpperCase() ??
      parsed['Category-Power']?.toUpperCase(),
    logAssisted:
      (parsed['CATEGORY-ASSISTED']?.toUpperCase() ?? false) === 'ASSISTED',
    logBand:
      parsed['CATEGORY-BAND']?.toUpperCase() ??
      parsed['Category-Band']?.toUpperCase(),
    logMode:
      parsed['CATEGORY-MODE']?.toUpperCase() ??
      parsed['Category-Mode']?.toUpperCase(),
    logFile: null
  };
};
