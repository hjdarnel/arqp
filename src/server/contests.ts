'use server';
import type { Contest } from '@prisma/client';
import { db } from './db';

export const getAllContests = async (): Promise<Contest[]> =>
  await db.contest
    .findMany({
      orderBy: { timeStart: 'desc' }
    })
    .catch((err) => {
      console.error('Error fetching all contests:', err);
      return [];
    });

export const getActiveContest = async (): Promise<Contest | null> =>
  await db.contest
    .findFirst({
      where: { timeStart: { lt: new Date() }, timeEnd: { gte: new Date() } }
    })
    .catch((err) => {
      console.error('Error fetching active contest:', err);
      return null;
    });

export const getLatestContest = async (): Promise<Contest | null> =>
  await db.contest
    .findFirst({
      where: {
        timeStart: { lt: new Date() }
      },
      orderBy: { timeEnd: 'desc' },
      take: 1
    })
    .catch((err) => {
      console.error('Error fetching latest contest:', err);
      return null;
    });
