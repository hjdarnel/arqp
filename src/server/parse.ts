import type { AuditLog, Contest } from '@prisma/client';
import { cabrilloToObject } from 'cabrillo';
import type { Category } from '~/util/categories';
import type { Location } from '~/util/locations';

export const parse = async (
  body: {
    callsign: string;
    email: string;
    score: number;
    category: Category;
    location: Location;
    multipleOperators: boolean;
    file: Blob;
  },
  currentContest: Contest,
): Promise<Omit<AuditLog, 'id' | 'created_at'>> => {
  type CabrilloLog = {
    QSO?: Record<string, unknown>[];
    Qso?: Record<string, unknown>[];
    'Claimed-Score'?: string;
    'CLAIMED-SCORE'?: string;
    'CATEGORY-OPERATOR'?: string;
    'Category-Operator'?: string;
    'CATEGORY-STATION'?: string;
    'Category-Station'?: string;
    'CATEGORY-TRANSMITTER'?: string;
    'Category-Transmitter'?: string;
    'CATEGORY-POWER'?: string;
    'Category-Power'?: string;
    'CATEGORY-ASSISTED'?: string;
    'CATEGORY-BAND'?: string;
    'Category-Band'?: string;
    'CATEGORY-MODE'?: string;
    'Category-Mode'?: string;
    [key: string]: unknown;
  };

  let parsed: CabrilloLog;
  try {
    parsed = cabrilloToObject(await body.file.text(), {
      contest: 'NAQP',
    });
  } catch (err) {
    console.error(err);
    throw new Error('Cannot parse file as Cabrillo log. Is it a valid log?');
  }

  for (const [key, value] of Object.entries(parsed)) {
    if (typeof value === 'string') {
      parsed[key] = value.trim().replace('\r', '');
    } else parsed[key] = value;
  }
  console.log(JSON.stringify(parsed, null, 2));

  if (!parsed.QSO?.length && !parsed.Qso?.length) {
    throw new Error('Unable to parse QSO records from body');
  }
  console.log(`Parsed ${parsed.QSO?.length ?? parsed.Qso?.length} QSO records`);
  console.info({ ...parsed, QSO: 'omitted' });

  return {
    contestId: currentContest.id,
    email: body.email,
    callsign: body.callsign,
    category: body.category,
    contestLocation: body.location,
    multipleOperators: body.multipleOperators,
    name: parsed.NAME ?? parsed.Name,
    club: parsed.CLUB ?? parsed.Club,
    claimedScore: body.score,
    logScore: parsed['CLAIMED-SCORE']
      ? Number.parseInt(parsed['CLAIMED-SCORE'])
      : parsed['Claimed-Score']
        ? Number.parseInt(parsed['Claimed-Score'])
        : null,
    logLocation: parsed.LOCATION ?? parsed.Location,
    logEmail: parsed.EMAIL ?? parsed.Email,
    logOperator:
      parsed['CATEGORY-OPERATOR']?.split('-')[0]?.toUpperCase() ??
      parsed['Category-Operator']?.split('-')[0]?.toUpperCase(),
    logStation:
      parsed['CATEGORY-STATION']?.toUpperCase() ??
      parsed['Category-Station']?.toUpperCase(),
    logTransmitter:
      parsed['CATEGORY-TRANSMITTER']?.toUpperCase() ??
      parsed['Category-Transmitter']?.toUpperCase(),
    logPower:
      parsed['CATEGORY-POWER']?.toUpperCase() ??
      parsed['Category-Power']?.toUpperCase(),
    logAssisted:
      (parsed['CATEGORY-ASSISTED']?.toUpperCase() ?? false) === 'ASSISTED',
    logBand:
      parsed['CATEGORY-BAND']?.toUpperCase() ??
      parsed['Category-Band']?.toUpperCase(),
    logMode:
      parsed['CATEGORY-MODE']?.toUpperCase() ??
      parsed['Category-Mode']?.toUpperCase(),
  } as Omit<AuditLog, 'id' | 'created_at'>;
};
