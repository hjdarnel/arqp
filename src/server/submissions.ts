'use server';
import type { Submission } from '@prisma/client';
import type { Category } from '~/util/categories';
import type { Location } from '~/util/locations';
import { db } from './db';

export const getSubmission = async (
  contestId: string,
  callsign: string
): Promise<Submission[] | null> => {
  const alphanum = /^[a-zA-Z0-9]+$/;

  if (!contestId || !alphanum.test(contestId)) {
    throw new Error(`Invalid contest id ${contestId}`);
  }

  if (!callsign || !alphanum.test(callsign)) {
    throw new Error(`Invalid callsign ${callsign}`);
  }

  return await db.submission.findMany({
    where: {
      contestId,
      callsign: {
        contains: callsign,
        mode: 'insensitive'
      }
    }
  });
};

export const getLatestContestSubmissions = async ({
  category,
  location
}: {
  category?: Category;
  location?: Location;
}) => {
  try {
    const latestContest = await db.contest.findFirst({
      where: {
        timeStart: { lt: new Date() }
      },
      orderBy: { timeEnd: 'desc' },
      take: 1
    });

    if (!latestContest) {
      return [];
    }

    return await db.submission.findMany({
      where: {
        contestId: latestContest?.id,
        category,
        contestLocation: location
      }
    });
  } catch (err) {
    console.error('Error fetching latest contest submissions:', err);
    return [];
  }
};
