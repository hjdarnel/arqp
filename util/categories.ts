export enum Category {
  DX = 'DX',
  Digital = 'Digital',
  'Mobile Inside Arkansas' = 'Mobile Inside Arkansas',
  'Mobile Outside Arkansas' = 'Mobile Outside Arkansas',
  'Multi Operator' = 'Multi Operator',
  'Portable Inside Arkansas' = 'Portable Inside Arkansas',
  'Portable Outside Arkansas' = 'Portable Outside Arkansas',
  Rover = 'Rover',
  'Single Operator High Power, Mixed' = 'Single Operator High Power, Mixed',
  'Single Operator Low Power, Mixed' = 'Single Operator Low Power, Mixed',
  'Single Operator QRP, Mixed' = 'Single Operator QRP, Mixed',
  'Single Operator High Power, All CW' = 'Single Operator High Power, All CW',
  'Single Operator Low Power, All CW' = 'Single Operator Low Power, All CW',
  'Single Operator High Power, All SSB' = 'Single Operator High Power, All SSB',
  'Single Operator Low Power, All SSB' = 'Single Operator Low Power, All SSB',
}
