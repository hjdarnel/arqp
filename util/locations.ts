export enum Location {
  'AR - Arkansas' = 'AR - Arkansas',
  'AR - Ashley' = 'AR - Ashley',
  'AR - Baxter' = 'AR - Baxter',
  'AR - Benton' = 'AR - Benton',
  'AR - Boone' = 'AR - Boone',
  'AR - Bradley' = 'AR - Bradley',
  'AR - Calhoun' = 'AR - Calhoun',
  'AR - Carroll' = 'AR - Carroll',
  'AR - Chicot' = 'AR - Chicot',
  'AR - Clark' = 'AR - Clark',
  'AR - Clay' = 'AR - Clay',
  'AR - Cleburne' = 'AR - Cleburne',
  'AR - Cleveland' = 'AR - Cleveland',
  'AR - Columbia' = 'AR - Columbia',
  'AR - Conway' = 'AR - Conway',
  'AR - Craighead' = 'AR - Craighead',
  'AR - Crawford' = 'AR - Crawford',
  'AR - Crittenden' = 'AR - Crittenden',
  'AR - Cross' = 'AR - Cross',
  'AR - Dallas' = 'AR - Dallas',
  'AR - Desha' = 'AR - Desha',
  'AR - Drew' = 'AR - Drew',
  'AR - Faulkner' = 'AR - Faulkner',
  'AR - Franklin' = 'AR - Franklin',
  'AR - Fulton' = 'AR - Fulton',
  'AR - Garland' = 'AR - Garland',
  'AR - Grant' = 'AR - Grant',
  'AR - Greene' = 'AR - Greene',
  'AR - Hempstead' = 'AR - Hempstead',
  'AR - Hot Spring' = 'AR - Hot Spring',
  'AR - Howard' = 'AR - Howard',
  'AR - Independence' = 'AR - Independence',
  'AR - Izard' = 'AR - Izard',
  'AR - Jackson' = 'AR - Jackson',
  'AR - Jefferson' = 'AR - Jefferson',
  'AR - Johnson' = 'AR - Johnson',
  'AR - Lafayette' = 'AR - Lafayette',
  'AR - Lawrence' = 'AR - Lawrence',
  'AR - Lee' = 'AR - Lee',
  'AR - Lincoln' = 'AR - Lincoln',
  'AR - Little River' = 'AR - Little River',
  'AR - Logan' = 'AR - Logan',
  'AR - Lonoke' = 'AR - Lonoke',
  'AR - Madison' = 'AR - Madison',
  'AR - Marion' = 'AR - Marion',
  'AR - Miller' = 'AR - Miller',
  'AR - Mississippi' = 'AR - Mississippi',
  'AR - Monroe' = 'AR - Monroe',
  'AR - Montgomery' = 'AR - Montgomery',
  'AR - Nevada' = 'AR - Nevada',
  'AR - Newton' = 'AR - Newton',
  'AR - Ouachita' = 'AR - Ouachita',
  'AR - Perry' = 'AR - Perry',
  'AR - Phillips' = 'AR - Phillips',
  'AR - Pike' = 'AR - Pike',
  'AR - Poinsett' = 'AR - Poinsett',
  'AR - Polk' = 'AR - Polk',
  'AR - Pope' = 'AR - Pope',
  'AR - Prairie' = 'AR - Prairie',
  'AR - Pulaski' = 'AR - Pulaski',
  'AR - Randolph' = 'AR - Randolph',
  'AR - St. Francis' = 'AR - St. Francis',
  'AR - Saline' = 'AR - Saline',
  'AR - Scott' = 'AR - Scott',
  'AR - Searcy' = 'AR - Searcy',
  'AR - Sebastian' = 'AR - Sebastian',
  'AR - Sevier' = 'AR - Sevier',
  'AR - Sharp' = 'AR - Sharp',
  'AR - Stone' = 'AR - Stone',
  'AR - Union' = 'AR - Union',
  'AR - Van Buren' = 'AR - Van Buren',
  'AR - Washington' = 'AR - Washington',
  'AR - White' = 'AR - White',
  'AR - Woodruff' = 'AR - Woodruff',
  'AR - Yell' = 'AR - Yell',
  Alabama = 'Alabama',
  Alaska = 'Alaska',
  Arizona = 'Arizona',
  California = 'California',
  Colorado = 'Colorado',
  Connecticut = 'Connecticut',
  Delaware = 'Delaware',
  Florida = 'Florida',
  Georgia = 'Georgia',
  Hawaii = 'Hawaii',
  Idaho = 'Idaho',
  Illinois = 'Illinois',
  Indiana = 'Indiana',
  Iowa = 'Iowa',
  Kansas = 'Kansas',
  Kentucky = 'Kentucky',
  Louisiana = 'Louisiana',
  Maine = 'Maine',
  Maryland = 'Maryland',
  Massachusetts = 'Massachusetts',
  Michigan = 'Michigan',
  Minnesota = 'Minnesota',
  Mississippi = 'Mississippi',
  Missouri = 'Missouri',
  Montana = 'Montana',
  Nebraska = 'Nebraska',
  Nevada = 'Nevada',
  'New Hampshire' = 'New Hampshire',
  'New Jersey' = 'New Jersey',
  'New Mexico' = 'New Mexico',
  'New York' = 'New York',
  'North Carolina' = 'North Carolina',
  'North Dakota' = 'North Dakota',
  Ohio = 'Ohio',
  Oklahoma = 'Oklahoma',
  Oregon = 'Oregon',
  Pennsylvania = 'Pennsylvania',
  'Rhode Island' = 'Rhode Island',
  'South Carolina' = 'South Carolina',
  'South Dakota' = 'South Dakota',
  Tennessee = 'Tennessee',
  Texas = 'Texas',
  Utah = 'Utah',
  Vermont = 'Vermont',
  Virginia = 'Virginia',
  Washington = 'Washington',
  'West Virginia' = 'West Virginia',
  Wisconsin = 'Wisconsin',
  Wyoming = 'Wyoming',
  Newfoundland = 'Newfoundland',
  'Prince Edward Island' = 'Prince Edward Island',
  'Nova Scotia' = 'Nova Scotia',
  'New Brunswick' = 'New Brunswick',
  Quebec = 'Quebec',
  Ontario = 'Ontario',
  Manitoba = 'Manitoba',
  Saskatchewan = 'Saskatchewan',
  Alberta = 'Alberta',
  'British Columbia' = 'British Columbia',
  'Yukon-NWT' = 'Yukon-NWT',
  DX = 'DX',
}
