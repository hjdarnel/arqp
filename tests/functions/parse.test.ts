import fs from 'node:fs';
import path from 'node:path';
/* eslint-disable @typescript-eslint/no-floating-promises */
import type { Contest } from '@prisma/client';
import { parse } from '../../src/server/parse';
import { Category } from '../../src/util/categories';
import { Location } from '../../src/util/locations';

const body = {
  email: 'abc123@gmail.com',
  callsign: 'ABC123',
  category: Category['Single Operator Low Power, All SSB'],
  location: Location['AR - Baxter'],
  multipleOperators: false,
  name: 'Henry Darnell',
  club: 'Big Bois',
  score: 652,
  file: {},
};

describe('snapshot tests', () => {
  it('works on normal input', async () => {
    const buffer = fs.readFileSync(
      path.join(__dirname, '../fixtures/normal.log'),
    );
    const blob = new Blob([buffer]);

    const result = await parse(
      {
        ...body,
        file: blob,
      },
      {
        id: '123',
      } as Contest,
    );

    expect(result).toMatchSnapshot();
  });

  it('works on lowercase input', async () => {
    const buffer = fs.readFileSync(
      path.join(__dirname, '../fixtures/lower.log'),
    );
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on padded input', async () => {
    const buffer = fs.readFileSync(
      path.join(__dirname, '../fixtures/padded.log'),
    );
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 1', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/1.txt'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 2', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/2.log'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 3', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/3.log'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 4', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/4.cbr'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 5', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/5.txt'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 6', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/6.LOG'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 7', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/7.log'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 8', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/8.log'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 9', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/9.log'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 10', async () => {
    const buffer = fs.readFileSync(path.join(__dirname, '../fixtures/10.txt'));
    const blob = new Blob([buffer]);

    const result = await parse({ ...body, file: blob }, {
      id: '123',
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it("doesn't work for x-qso", async () => {
    const buffer = fs.readFileSync(
      path.join(__dirname, '../fixtures/x-qso.log'),
    );
    const blob = new Blob([buffer]);

    expect(
      parse({ ...body, file: blob }, {
        id: '123',
      } as Contest),
    ).rejects.toThrow('Unable to parse QSO records from body');
  });

  it(`doesn't fail`, () => {
    expect(true).toBe(true);
  });
});
