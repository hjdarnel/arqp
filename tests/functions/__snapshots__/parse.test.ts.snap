// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests works on 1 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "FLORIDA CONTEST GROUP",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": undefined,
  "logLocation": "NFL",
  "logMode": "MIXED",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": null,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on 2 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": undefined,
  "logLocation": "NC",
  "logMode": "SSB",
  "logOperator": "SINGLE",
  "logPower": "HIGH",
  "logScore": 211,
  "logStation": "",
  "logTransmitter": "",
  "multipleOperators": false,
  "name": "REDACT",
}
`;

exports[`snapshot tests works on 3 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "Arizona Outlaws Contest Club",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": true,
  "logBand": "ALL",
  "logEmail": "REDACT@outlook.com",
  "logLocation": "AZ",
  "logMode": "SSB",
  "logOperator": undefined,
  "logPower": "HIGH",
  "logScore": 299,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Sheila Arlene Schwartz",
}
`;

exports[`snapshot tests works on 4 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": undefined,
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": undefined,
  "logEmail": undefined,
  "logLocation": undefined,
  "logMode": undefined,
  "logOperator": undefined,
  "logPower": undefined,
  "logScore": null,
  "logStation": undefined,
  "logTransmitter": undefined,
  "multipleOperators": false,
  "name": undefined,
}
`;

exports[`snapshot tests works on 5 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": "REDACT@gmail.com",
  "logLocation": "",
  "logMode": "MX",
  "logOperator": "SINGLE",
  "logPower": "LOW",
  "logScore": null,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "",
}
`;

exports[`snapshot tests works on 6 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "Hudson Valley Contesters and DXers",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "40M",
  "logEmail": "REDACT@AOL.COM",
  "logLocation": "ENY",
  "logMode": "SSB",
  "logOperator": undefined,
  "logPower": "LOW",
  "logScore": 16,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on 7 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": true,
  "logBand": "ALL",
  "logEmail": "REDACT@COMCAST.NET",
  "logLocation": "IL",
  "logMode": "CW",
  "logOperator": "SINGLE",
  "logPower": "HIGH",
  "logScore": 505,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on 8 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "Snohomish County Hams Club",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": "REDACT@frontier.com",
  "logLocation": "WA",
  "logMode": "SSB",
  "logOperator": "REDACT",
  "logPower": "LOW",
  "logScore": 9,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on 9 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "Tennessee Contest Group",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": true,
  "logBand": "ALL",
  "logEmail": "REDACT@ARRL.NET",
  "logLocation": "TN",
  "logMode": "SSB",
  "logOperator": "REDACT",
  "logPower": "LOW",
  "logScore": 204,
  "logStation": "FIXED",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on 10 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "NONE",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": undefined,
  "logEmail": "REDEACT@hotmail.com",
  "logLocation": undefined,
  "logMode": undefined,
  "logOperator": undefined,
  "logPower": undefined,
  "logScore": 790,
  "logStation": undefined,
  "logTransmitter": undefined,
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on lowercase input 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": undefined,
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": "ABC123@arrl.net",
  "logLocation": "AR",
  "logMode": "SSB",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": 652,
  "logStation": "PORTABLE",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on normal input 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": undefined,
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": "ABC123@arrl.net",
  "logLocation": "AR",
  "logMode": "SSB",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": 652,
  "logStation": "PORTABLE",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on padded input 1`] = `
{
  "callsign": "ABC123",
  "category": "Single Operator Low Power, All SSB",
  "claimedScore": 652,
  "club": "Hudson",
  "contestId": "123",
  "contestLocation": "AR - Baxter",
  "email": "abc123@gmail.com",
  "logAssisted": true,
  "logBand": undefined,
  "logEmail": "REDACT.qrp@gmail.com",
  "logLocation": "NY",
  "logMode": "MIXED",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": 1200,
  "logStation": "FIXED",
  "logTransmitter": undefined,
  "multipleOperators": false,
  "name": "REDACT",
}
`;
